node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,513 @@
+			job.run_loop().await
+		}
+		.boxed()
+	}
+}
+
+impl ProvisioningJob {
+	pub fn new(
+		relay_parent: Hash,
+		sender: mpsc::Sender<FromJob>,
+		receiver: mpsc::Receiver<ToJob>,
+	) -> Self {
+		Self {
+			relay_parent,
+			sender,
+			receiver,
+			provisionable_data_channels: Vec::new(),
+			backed_candidates: Vec::new(),
+			signed_bitfields: Vec::new(),
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			use ProvisionerMessage::{
+				ProvisionableData, RequestBlockAuthorshipData, RequestInherentData,
+			};
+
+			match msg {
+				ToJob::Provisioner(RequestInherentData(_, sender)) => {
+					send_inherent_data(
+						self.relay_parent,
+						&self.signed_bitfields,
+						&self.backed_candidates,
+						sender,
+						self.sender.clone(),
+					)
+					.await?
+				}
+				ToJob::Provisioner(RequestBlockAuthorshipData(_, sender)) => {
+					self.provisionable_data_channels.push(sender)
+				}
+				ToJob::Provisioner(ProvisionableData(data)) => {
+					let mut bad_indices = Vec::new();
+					for (idx, channel) in self.provisionable_data_channels.iter_mut().enumerate() {
+						match channel.send(data.clone()).await {
+							Ok(_) => {}
+							Err(_) => bad_indices.push(idx),
+						}
+					}
+					self.note_provisionable_data(data);
+
+					// clean up our list of channels by removing the bad indices
+					// start by reversing it for efficient pop
+					bad_indices.reverse();
+					// Vec::retain would be nicer here, but it doesn't provide
+					// an easy API for retaining by index, so we re-collect instead.
+					self.provisionable_data_channels = self
+						.provisionable_data_channels
+						.into_iter()
+						.enumerate()
+						.filter(|(idx, _)| {
+							if bad_indices.is_empty() {
+								return true;
+							}
+							let tail = bad_indices[bad_indices.len() - 1];
+							let retain = *idx != tail;
+							if *idx >= tail {
+								bad_indices.pop();
+							}
+							retain
+						})
+						.map(|(_, item)| item)
+						.collect();
+				}
+				ToJob::Stop => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	fn note_provisionable_data(&mut self, provisionable_data: ProvisionableData) {
+		match provisionable_data {
+			ProvisionableData::Bitfield(_, signed_bitfield) => {
+				self.signed_bitfields.push(signed_bitfield)
+			}
+			ProvisionableData::BackedCandidate(backed_candidate) => {
+				self.backed_candidates.push(backed_candidate)
+			}
+			_ => {}
+		}
+	}
+}
+
+type CoreAvailability = BitVec<bitvec::order::Lsb0, u8>;
+
+// preprocessing the cores involves a bit more data than is comfortable in a tuple, so let's make a struct of it
+struct PreprocessedCore {
+	assumption: OccupiedCoreAssumption,
+	scheduled_core: ScheduledCore,
+	availability: Option<CoreAvailability>,
+	timeout: Option<BlockNumber>,
+	idx: usize,
+}
+
+impl PreprocessedCore {
+	fn new(idx: usize, core: CoreState) -> Option<Self> {
```

Comment:
I accidentally made my other comment on a commit instead of the full diff so it shows up badly here.

Reposting: 

My post

> The preprocessing is supposed to be this:
>  1. Given the initial core state as given by the runtime API
>  1. And the bitfields that we have accumulated off-chain
>  1. What is the actual state of the core?
>
> So it's not about "did the initial state say this was next up" but "did the initial state say this was next up and was the condition for next_up met"?. This question can't be answered without looking at the set of given bitfields.

@coriolinus 
> PreprocessedCore just collects some data; the bitfields are checked in PreprocessedCore::choose_candidate.

Reply:
My point is that this is selecting the wrong data and that `choose_candidate` is the wrong place to be checking the bitfields.

Imagine you have a candidate which has `next_up_on_available = Some(para_a)` and `next_up_on_timeout = `Some(para_b)`. This code will select an `OccupiedCoreAssumption::Available`, although it hasn't checked against the set of bitfields to see if the core actually has become available.

Then, in `choose_candidate`, if the bitfields show that the core did not become available, no candidate will be selected.

----

To be fair, this bug doesn't ever seem to cause the node to include things that shouldn't be included, but it does cause the node to not include things that should be included.

Reply:
https://github.com/paritytech/polkadot/pull/1473/commits/0e97e0108fa6b0f245d770ad2874ccba778fe746